### PR TITLE
Safari key window AIS-122

### PR DESCRIPTION
**Pull request by E-B-Smith**
Creating a SFSafariViewController and adding a temporary key window was interfering with partners applications:

* Losing the keyWindow focus stalled UI initialization.
* Controls would lose firstResponder status.
* The Facebook login controller wouldn't appear correctly.

This patch fixes those problems.

### CODE REVIEW

**Comment by aaustin** [comment]
Again, make sure you tag someone to review this or it's just going to sit around. I'll review now.

File: Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m [hunk 1]
```diff
@@ -6,36 +6,95 @@
...
-NSInteger const ABOUT_30_DAYS_TIME_IN_SECONDS = 60 * 60 * 24 * 30;
+#else   // ------------------------------------------------------------------------------ iOS >= 9.0
 
-@interface BNCStrongMatchHelper ()
 
-@property (strong, nonatomic) UIWindow *secondWindow;
+#pragma mark - BNCMatchViewController
+
+
+#import <SafariServices/SafariServices.h>
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m]: Why is this included? You can't just add a build-level requirement of SafariServices in a new release. That'll break thousands of builds when they upgrade if they hadn't included it to begin with...

File: Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m [hunk 2]
```diff
@@ -95,69 +155,121 @@ + (NSURL *)getUrlForCookieBasedMatchingWithBranchKey:(NSString *)branchKey redir
...
+    Class BNCMatchViewControllerSubclass = NSClassFromString(@"BNCMatchViewController_Safari");
+    if (!BNCMatchViewControllerSubclass) {
+        //  The class isn't registered.  Create it:
+
+        BNCMatchViewControllerSubclass =
+            objc_allocateClassPair(SFSafariViewControllerClass, "BNCMatchViewController_Safari", 0);
+        if (!BNCMatchViewControllerSubclass) return NO;
+
+        BOOL fail = NO;
+        fail |= ![self subclass:BNCMatchViewControllerSubclass selector:@selector(becomeFirstResponder)];
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m]: Why are you hooking these up? The methods don't do anything? Maybe this PR isn't ready for review?

File: Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m (hunk 1 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m]: I import the header file but don't require that the library be linked to the app.
I tested this on Xcode 7 and Xcode 8, with and without SafariServices linked.

File: Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m (hunk 2 above)
Comment (E-B-Smith) [inline Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m]: These methods refuse firstResponder status so that the SFSafariViewController view doesn't steal the focus.

**Comment by E-B-Smith** [comment]
I dynamically subclassed the SFSafariViewController so that it would not become the first responder, and so it wouldn't steal the focus from other views or controls.

I do this dynamically at runtime so that we don't have to require that the SafariServices library is linked to the app.

This PR was tested on Xcode 7 and 8, iOS 7/8/9/10.  I tested with both SafariServices linked and not linked.

File: Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m (hunk 2 above)
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m]: Ahh interesting.

File: Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m [hunk 3]
```diff
@@ -95,69 +155,121 @@ + (NSURL *)getUrlForCookieBasedMatchingWithBranchKey:(NSString *)branchKey redir
...
+     self.primaryWindow = nil;
+
+    [BNCPreferenceHelper preferenceHelper].lastStrongMatchDate = [NSDate date];
+    self.shouldDelayInstallRequest = NO;
+    self.requestInProgress = NO;
 }
 
+- (void)safariViewController:(SFSafariViewController *)controller
+      didCompleteInitialLoad:(BOOL)didLoadSuccessfully {
+    NSLog(@"Safari Did load. Success: %d.", didLoadSuccessfully);
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m]: The problem is that this is only place you're referencing a class that can't exist when SafariServices.framework is not present. I saw your note that you tested without SafariServices.framework, but I just don't understand how this could technically build without the class present?

File: Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m [hunk 4]
```diff
@@ -6,36 +6,95 @@
...
+
+@implementation BNCMatchViewController
+
+- (instancetype) initWithURL:(NSURL*)URL {
+    self = [super init];
+    return self;
+}
+
+- (BOOL) canBecomeFirstResponder {
+    //NSLog(@"Can be!!! Responder.");
```
Comment (aaustin) [inline Branch-SDK/Branch-SDK/BNCStrongMatchHelper.m]: Just for cleanliness - can you remove this logs?

**Comment by aaustin** [comment]
@derrickstaten It'd be good for you to take a gander at this PR too since you're most familiar with this code.

@E-B-Smith Have you tested that before /v1/install, there's a request to /strong_match? And that this functionality still works as intended?

File: Branch-TestBed/Branch-TestBed/Base.lproj/Main.storyboard [hunk 5]
```diff
@@ -1,13 +1,10 @@
 <?xml version="1.0" encoding="UTF-8" standalone="no"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="11542" systemVersion="15G1108" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" colorMatched="YES" initialViewController="rgL-wI-yV3">
```
Comment (derrickstaten) [inline Branch-TestBed/Branch-TestBed/Base.lproj/Main.storyboard]: Are all of these changes to the storyboard required? If not, I vote for removing this from the PR

**Comment by derrickstaten** [comment]
If this (1) works on iOS 10 (and preferably 9) and does not cause issues compiling on older Xcodes and OSes, then this looks fine to me.

**Comment by E-B-Smith** [comment]
@derrickstaten  Yes, I tested this code on iOS 7/8/9/10 and Xcode 7 & 8.  That's why the storyboards were updated.
